### PR TITLE
[65183] sam0: initialize GCLK2 in wdt_sam0_init

### LINKED ISSUE #65183: samd21: wdt clock is not initialized correctly
**Describe the bug**
Watchdog Implementation for Microchip SAMD21 is missing clock initialization. [wdt_sam0_init()](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/watchdog/wdt_sam0.c#L266) assumes GCLK2 to be a 1.024kHz clock, however GCLK2 in not initialized anywhere in the code, which results in unpredictable watchdog behavior. The solution is to properly configure GCLK2 before enabling WDT:
``` diff
static int wdt_sam0_init(const struct device *dev)
{
#ifdef CONFIG_WDT_DISABLE_AT_BOOT
    /* Ignore any errors */
    wdt_sam0_disable(dev);
#endif
    /* Enable APB clock */
#ifdef MCLK
    MCLK->APBAMASK.bit.WDT_ = 1;

    /* watchdog clock is fed by OSCULP32K */
#else
    PM->APBAMASK.bit.WDT_ = 1;

+    /* Enable GCLK to output OSCULP32K divided by 32 */
+    GCLK->GENDIV.reg = GCLK_GENDIV_ID(2) | GCLK_GENDIV_DIV(4);
+    GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(2) | GCLK_GENCTRL_GENEN |
+                        GCLK_GENCTRL_SRC_OSCULP32K | GCLK_GENCTRL_DIVSEL;
+    while (GCLK->STATUS.bit.SYNCBUSY) ;

    /* Connect to GCLK2 (~1.024 kHz) */
    GCLK->CLKCTRL.reg = GCLK_CLKCTRL_ID_WDT
        | GCLK_CLKCTRL_GEN_GCLK2
        | GCLK_CLKCTRL_CLKEN;
#endif

    IRQ_CONNECT(DT_INST_IRQN(0),
            DT_INST_IRQ(0, priority), wdt_sam0_isr,
            DEVICE_DT_INST_GET(0), 0);
    irq_enable(DT_INST_IRQN(0));

    return 0;
}
```

**To Reproduce**
1. configure WDT
2. observe WDT timeout not matching requested timeout

**Impact**
This renders watchdog on SAMD21 unusable

### PR DESCRIPTION
**Pull request by ghost**
Initialize GCLK2 to output 1.024kHz required by watchdog timer.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/65183

### CODE REVIEW

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Thanks! Please reformat the Git commit log to follow the project standard (please see https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) and add a DCO sign-off (see https://docs.zephyrproject.org/latest/contribute/guidelines.html#developer-certification-of-origin-dco).

**Comment by attie-argentum** [comment]
Thanks for the patch!

This would probably fit better in [`soc/arm/atmel_sam0/common/soc_samd2x.c`](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/atmel_sam0/common/soc_samd2x.c), where GCLK Gen 2 is indeed marked up for the WDT.

**Comment by nandojve** [comment]
Hi @vlm-laba7 ,

Please, rebase on top of main and make sure CI is green.

@attie-argentum , could you revisit this?

**Review by attie-argentum — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

**Comment by nandojve** [comment]
Hi @vlm-laba7 ,

Make sure you fix compliance check otherwise it is not possible to merge your PR:

```
Error: See https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-guidelines for more details

 d5b2a263d918263113ce3203a5291c5b04f0bfbd: author email (vlm <vlad@laba7.com>) needs to match one of the signed-off-by entries.
d5b2a[26](https://github.com/zephyrproject-rtos/zephyr/actions/runs/6877251017/job/18754400492?pr=65211#step:9:27)3d918263113ce3203a5291c5b04f0bfbd: author email (vlm <vlad@laba7.com>) does not follow the syntax: First Last <email>.
```

File: soc/arm/atmel_sam0/common/soc_samd2x.c [hunk 1]
```diff
@@ -246,6 +246,20 @@ static inline void gclk_adc_configure(void)
...
+{
+	GCLK->GENDIV.reg = GCLK_GENDIV_ID(2) | GCLK_GENDIV_DIV(4);
+	GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(2) | GCLK_GENCTRL_GENEN |
+						GCLK_GENCTRL_SRC_OSCULP32K | GCLK_GENCTRL_DIVSEL;
+
+	while (GCLK->STATUS.bit.SYNCBUSY) {
+	}
+}
+#endif
+
```
Comment (nandojve) [inline soc/arm/atmel_sam0/common/soc_samd2x.c:262]: I think positive logic is easy to read. BTW, you need align correct as below:

```c
#if CONFIG_WDT_SAM0
static inline void gclk_wdt_configure(void)
{
	GCLK->GENDIV.reg = GCLK_GENDIV_ID(2)
			 | GCLK_GENDIV_DIV(4);

	GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(2)
			  | GCLK_GENCTRL_GENEN
			  | GCLK_GENCTRL_SRC_OSCULP32K
			  | GCLK_GENCTRL_DIVSEL;

	while (GCLK->STATUS.bit.SYNCBUSY) {
	}
}
#else
#define gclk_wdt_configure()
#endif
```

I was wondering if below is more readable

```c
static inline void gclk_wdt_configure(void)
{
	if (IS_ENABLED(CONFIG_WDT_SAM0)) {
		GCLK->GENDIV.reg = GCLK_GENDIV_ID(2)
				 | GCLK_GENDIV_DIV(4);

		GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(2)
				  | GCLK_GENCTRL_GENEN
				  | GCLK_GENCTRL_SRC_OSCULP32K
				  | GCLK_GENCTRL_DIVSEL;

		while (GCLK->STATUS.bit.SYNCBUSY) {
		}
	}
}
```

**Review by nandojve — CHANGES_REQUESTED** [review]

**Review by nandojve — APPROVED** [review]

**Review by pdgendt — APPROVED** [review]

**Comment by stephanosio** [comment]
> Signed-off-by: Vlad Laba7 <vlad@laba7.com>

@vlm-laba7 Unless your last name actually contains `7`, please use your full name as required by the DCO policy.

**Comment by nandojve** [comment]
Hi @kartben,

> @stephanosio sorry, I prefer not to use any names on this account - feel free to author this commit

Do you know who can help to clarify this?

**Comment by kartben** [comment]
> feel free to author this commit

Unfortunately I don't think we can accept contributions not covered by the DCO, as that's how the contributor acknowledges they have the rights to contribute the code to Zephyr under the Apache 2.0 License.

**Comment by nandojve** [comment]
> > feel free to author this commit
> 
> Unfortunately I don't think we can accept contributions not covered by the DCO, as that's how the contributor acknowledges they have the rights to contribute the code to Zephyr under the Apache 2.0 License.

This means that someone should open another PR with a solution and we need to close this, right?

**Review by attie-argentum — APPROVED** [review]
Untested (I don't have the hardware), but looks sensible to me - thanks!

**Review by henrikbrixandersen — APPROVED** [review]

**Pull request merged**